## 46-881 Class 6

### Python pandas: Data Exploration and Wrangling (Part 2)

Amr Farahat

CMU / Tepper

2022-12-05

---

## Exploratory Data Analysis using Python Exercise: PIT Flight Departures - SOLUTIONS

Data source: Bureau of Transportation Statistics.
Covers domestic flights only
https://www.transtats.bts.gov/DL_SelectFields.aspx?gnoyr_VQ=FGJ&QO_fu146_anzr=b0-gvzr

The database contains scheduled and actual departure and arrival times reported by certified U.S. air carriers that account for at least one percent of domestic scheduled passenger revenues. The data is collected by the Office of Airline Information, Bureau of Transportation Statistics (BTS).

Use the exercise as an opportunity to review / reference / get familiar with Python's pandas package. References:
- <https://wesmckinney.com/book/>
- <https://pandas.pydata.org/>

### Section 0: Setting up

**Task 0.1**: Import the numpy and pandas libraries (also called packages or modules) 

In [23]:
import numpy as np
import pandas as pd

**Task 0.2**: Load the dataset "PIT_2021.csv" into a dataframe labelled "df"

In [24]:
df = pd.read_csv("PIT_2021.csv")

**Task 0.3**: Run the code cell below to disable limits on the number of columns displayed in Jupyter (so you can scroll across and inspect all columns)  

In [25]:
pd.options.display.max_columns = None
pd.options.display.max_rows = 20

### Section 1: First glance at the data 
*To check its structure and that it was loaded properly...*

**Task 1.1**: Find the dimensions of "df"; that is, the number of rows and number of columns. 

In [26]:
df.shape

(32620, 33)

*We have 32620 fights, described by 33 columns, flown out of PIT in 2021. These are all the domestic flights flown out of PIT by certified U.S. carriers with at least 1% of domestic scheduled passenger revenue.*

**Task 1.2**: Display the list of column names of df.

In [27]:
print(list(df.columns))

['YEAR', 'MONTH', 'DAY_OF_MONTH', 'DAY_OF_WEEK', 'FL_DATE', 'OP_CARRIER', 'OP_CARRIER_FL_NUM', 'ORIGIN', 'ORIGIN_CITY_NAME', 'ORIGIN_STATE_ABR', 'DEST', 'DEST_CITY_NAME', 'DEST_STATE_ABR', 'CRS_DEP_TIME', 'DEP_TIME', 'DEP_DELAY', 'DEP_DELAY_NEW', 'DEP_DEL15', 'DEP_TIME_BLK', 'CRS_ARR_TIME', 'ARR_TIME', 'ARR_DELAY', 'ARR_DELAY_NEW', 'ARR_DEL15', 'ARR_TIME_BLK', 'CANCELLED', 'CANCELLATION_CODE', 'DIVERTED', 'CARRIER_DELAY', 'WEATHER_DELAY', 'NAS_DELAY', 'SECURITY_DELAY', 'LATE_AIRCRAFT_DELAY']


**Task 1.3**: Inspect the first 5 rows and last 5 rows of df. Note the default index column 0 .. length-1.

In [28]:
df.head(5) #or df.head() because default is 5 rows

,YEAR,MONTH,DAY_OF_MONTH,DAY_OF_WEEK,FL_DATE,OP_CARRIER,OP_CARRIER_FL_NUM,ORIGIN,ORIGIN_CITY_NAME,ORIGIN_STATE_ABR,DEST,DEST_CITY_NAME,DEST_STATE_ABR,CRS_DEP_TIME,DEP_TIME,DEP_DELAY,DEP_DELAY_NEW,DEP_DEL15,DEP_TIME_BLK,CRS_ARR_TIME,ARR_TIME,ARR_DELAY,ARR_DELAY_NEW,ARR_DEL15,ARR_TIME_BLK,CANCELLED,CANCELLATION_CODE,DIVERTED,CARRIER_DELAY,WEATHER_DELAY,NAS_DELAY,SECURITY_DELAY,LATE_AIRCRAFT_DELAY
0,2021,1,10,7,1/10/2021 12:00:00 AM,OO,3878,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,1901,1900.0,-1.0,0.0,0.0,1900-1959,2025,2023.0,-2.0,0.0,0.0,2000-2059,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
1,2021,1,4,1,1/4/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,756.0,-4.0,0.0,0.0,0800-0859,921,904.0,-17.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
2,2021,1,3,7,1/3/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,754.0,-6.0,0.0,0.0,0800-0859,921,903.0,-18.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
3,2021,1,3,7,1/3/2021 12:00:00 AM,OO,3989,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,1245,1243.0,-2.0,0.0,0.0,1200-1259,1403,1339.0,-24.0,0.0,0.0,1400-1459,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
4,2021,1,11,1,1/11/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,807.0,7.0,7.0,0.0,0800-0859,930,913.0,-17.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN


In [29]:
df.tail(5) #or df.tail() because default is 5 rows

,YEAR,MONTH,DAY_OF_MONTH,DAY_OF_WEEK,FL_DATE,OP_CARRIER,OP_CARRIER_FL_NUM,ORIGIN,ORIGIN_CITY_NAME,ORIGIN_STATE_ABR,DEST,DEST_CITY_NAME,DEST_STATE_ABR,CRS_DEP_TIME,DEP_TIME,DEP_DELAY,DEP_DELAY_NEW,DEP_DEL15,DEP_TIME_BLK,CRS_ARR_TIME,ARR_TIME,ARR_DELAY,ARR_DELAY_NEW,ARR_DEL15,ARR_TIME_BLK,CANCELLED,CANCELLATION_CODE,DIVERTED,CARRIER_DELAY,WEATHER_DELAY,NAS_DELAY,SECURITY_DELAY,LATE_AIRCRAFT_DELAY
32615,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2407,PIT,"Pittsburgh, PA",PA,IAH,"Houston, TX",TX,1639,1645.0,6.0,6.0,0.0,1600-1659,1859,1853.0,-6.0,0.0,0.0,1800-1859,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32616,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2109,PIT,"Pittsburgh, PA",PA,DEN,"Denver, CO",CO,838,833.0,-5.0,0.0,0.0,0800-0859,1022,1019.0,-3.0,0.0,0.0,1000-1059,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32617,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2108,PIT,"Pittsburgh, PA",PA,IAH,"Houston, TX",TX,610,604.0,-6.0,0.0,0.0,0600-0659,835,822.0,-13.0,0.0,0.0,0800-0859,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32618,2021,12,20,1,12/20/2021 12:00:00 AM,UA,1810,PIT,"Pittsburgh, PA",PA,DEN,"Denver, CO",CO,1600,1704.0,64.0,64.0,1.0,1600-1659,1744,1841.0,57.0,57.0,1.0,1700-1759,0.0,NaN,0.0,57.0,0.0,0.0,0.0,0.0
32619,2021,12,20,1,12/20/2021 12:00:00 AM,UA,1226,PIT,"Pittsburgh, PA",PA,EWR,"Newark, NJ",NJ,1635,1630.0,-5.0,0.0,0.0,1600-1659,1802,1812.0,10.0,10.0,0.0,1800-1859,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN


**Task 1.4**: Run the info() method on df and observe the information on missing data.

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32620 entries, 0 to 32619
Data columns (total 33 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   YEAR                 32620 non-null  int64  
 1   MONTH                32620 non-null  int64  
 2   DAY_OF_MONTH         32620 non-null  int64  
 3   DAY_OF_WEEK          32620 non-null  int64  
 4   FL_DATE              32620 non-null  object 
 5   OP_CARRIER           32620 non-null  object 
 6   OP_CARRIER_FL_NUM    32620 non-null  int64  
 7   ORIGIN               32620 non-null  object 
 8   ORIGIN_CITY_NAME     32620 non-null  object 
 9   ORIGIN_STATE_ABR     32620 non-null  object 
 10  DEST                 32620 non-null  object 
 11  DEST_CITY_NAME       32620 non-null  object 
 12  DEST_STATE_ABR       32620 non-null  object 
 13  CRS_DEP_TIME         32620 non-null  int64  
 14  DEP_TIME             32177 non-null  float64
 15  DEP_DELAY            32177 non-null 

*Departure and Arrival time information and delay is missing some variables. This can largely be explained by cancelled and diverted flights.*

### Section 2: Exploring Departure Time Patterns

**Task 2.1**: Compute the number of flights departing each day of the week (Monday through Sunday). What do you observe? Note (a): Monday=1, Sunday=7; Note (b): 2021 started on a Friday and ended on a Friday (so one more Friday in the year than the other week days). On average, which day(s) of the week are busiest in terms of departures.

In [9]:
#df["DAY_OF_WEEK"].value_counts().sort_index()
df.pivot_table(index="DAY_OF_WEEK", values="OP_CARRIER", aggfunc="count")

,OP_CARRIER
DAY_OF_WEEK,
1,4898
2,4191
3,4246
4,5055
5,4930
6,4297
7,5003


*Sundays, Mondays, Thursdays, and Fridays are the busiest days. This makes sense given the typical combined pattern of personal and business travel. The Friday number should be adjust by the number of days in the year 4930 * 52/53 = 4837.*

**Task 2.2**: Compute the number of flights scheduled to depart each hour of the day. Hint: Use "DEP_TIME_BLK". What do you observe? 

In [22]:
#df["DEP_TIME_BLK"].value_counts().sort_index()
df.pivot_table(index="DEP_TIME_BLK", values="OP_CARRIER", aggfunc="count")
#df.pivot_table(index=pd.to_datetime(df["CRS_DEP_TIME"], format= "%H%m").dt.hour, values="OP_CARRIER", aggfunc="count")

,OP_CARRIER
DEP_TIME_BLK,
0001-0559,630
0600-0659,4778
0700-0759,3336
0800-0859,2238
0900-0959,1114
1000-1059,1917
1100-1159,2399
1200-1259,2162
1300-1359,1916


*Early morning hours 6 a.m. - 8 a.m. have significantly more departures than any other hour of the day. Yes, allow for longer security line delays!*

**Task 2.3**: Compute the fraction of flights whose departure is delayed 15 or more mins for each hour of the day. What do you observe? 

In [23]:
#df["DEP_DEL15"].groupby(df["DEP_TIME_BLK"]).mean()
df.pivot_table(index="DEP_TIME_BLK", values="DEP_DEL15", aggfunc="mean")

,DEP_DEL15
DEP_TIME_BLK,
0001-0559,0.054226
0600-0659,0.056847
0700-0759,0.058519
0800-0859,0.067574
0900-0959,0.077768
1000-1059,0.074948
1100-1159,0.114346
1200-1259,0.141050
1300-1359,0.130458


*System delays accummulate throughout the day...*

### Section 3: Exploring Carrier (Airline) Patterns

**Task 3.1**: How many airlines depart from PIT? Take a moment to familiarize yourself with these airline codes in the accompanying file "Carrier_codes.xlsx". 

In [13]:
df["OP_CARRIER"].unique()

array(['OO', 'UA', 'WN', 'YV', 'YX', '9E', 'AA', 'AS', 'B6', 'DL', 'F9',
       'G4', 'MQ', 'NK', 'OH'], dtype=object)

In [14]:
len(df["OP_CARRIER"].unique())

15

*15 airlines (at least these are the main ones required to report to the Beureau of Transportation Statistics). The airlines are: Alaska (AS), American (AA), Allegiant (G4), Delta (DL), Endeavor (9E), Envoy (MQ), Frontier (F9), JetBlue (B6), Mesa (YV), PSA (OH), Republic (YX), SkyWest (OO), Southwest (WN), Spirit (NK), and United (UA). Some of these are regional airlines operating flight for larger airlines.* 

**Task 3.2**: Compute the number of departing flights flown by each airline. Do you find anything surprising in this summary? 

In [30]:
#by_carrier = df.groupby("OP_CARRIER")
#by_carrier.size().sort_values(ascending = False)
df.pivot_table(index="OP_CARRIER", values="ORIGIN", aggfunc="count").sort_values(by="ORIGIN", ascending=False)

,ORIGIN
OP_CARRIER,
YX,8828
WN,7061
AA,3140
NK,2205
OO,2158
DL,2119
G4,1325
UA,1317
YV,949


*Republic Airlines has the most departures out of PIT. It is a regional airline serving American, Delta, and United. Following Republic is Southwest Airlines.* 

**Task 3.3**: Compute the fraction of flights whose departure is delayed 15 or more mins for each airline operating out of PIT. Which airline(s) appear to have the best on-time performance? Which airline(s) appear to have the worst on-time performance?

In [32]:
#df["DEP_DEL15"].groupby(df["OP_CARRIER"]).mean().sort_values(ascending=False)
df.pivot_table(index="OP_CARRIER", values="DEP_DEL15", aggfunc="mean").sort_values(by="DEP_DEL15")

,DEP_DEL15
OP_CARRIER,
DL,0.069635
YX,0.079652
9E,0.088475
UA,0.088889
YV,0.114347
AA,0.140777
OO,0.155691
B6,0.156648
OH,0.157303


*By that metric, Delta, Republic, Endeavor, and United have the best average performance while Frontier, Allegiant, Alaska, and Envoy have the worst.*

### Section 4: A More Detailed Look at Airline Delay / On-Time Statistics

**Task 4.1**: Create a new columns in "df": "ARR_ON_TIME". A flight is on time if it's not cancelled/diverted *and* arrives earlier than 15 minutes past its scheduled arrival time. 

In [33]:
df["ARR_ON_TIME"] = ((df["ARR_DEL15"] == 0) & (df["CANCELLED"] == 0) & (df["DIVERTED"] == 0)).astype(int)

**Task 4.2**: Create a table comparing all airlines in terms of these performance metrics (using the average of the columns). Which airline(s) have the best on-time performance?

In [36]:
perf_cols = ["ARR_DELAY", "ARR_DELAY_NEW", "ARR_DEL15", "ARR_ON_TIME"]

In [37]:
#df[perf_cols].groupby(df["OP_CARRIER"]).mean().sort_values("DEP_DEL15", ascending=False)
df.pivot_table(index="OP_CARRIER", values=perf_cols, aggfunc="mean").sort_values(by="ARR_ON_TIME", ascending=False)

,ARR_DEL15,ARR_DELAY,ARR_DELAY_NEW,ARR_ON_TIME
OP_CARRIER,,,,
DL,0.067299,-5.345972,5.573934,0.928740
9E,0.095794,-3.261682,10.795561,0.901048
YX,0.095331,-4.721579,6.451876,0.893294
UA,0.118098,-1.454755,8.874233,0.873197
AA,0.121626,-2.639675,11.132033,0.860191
YV,0.133261,0.472373,14.645720,0.842993
OH,0.152370,2.193002,13.138826,0.839106
B6,0.156934,3.408759,14.182482,0.832432
OO,0.155587,4.819425,15.141914,0.829935


*Delta's performance beats any other airline across all metrics*

*Compared to American Airlines, Southwest is doing worse on ARR_DEL15 and ARR_ON_TIME but better on ARR_DELAY_NEW (minutes) which suggests that when they're delayed they're not delayed by much.*

**Task 4.3**: Examine if these statistics are comparable by looking at patterns *where* airlines fly.

In [45]:
#pd.crosstab(df["OP_CARRIER"], df["DEST_STATE_ABR"], margins=True, normalize="index")
pd.options.display.max_rows = 20
#df.pivot_table(index=["OP_CARRIER","DEST_STATE_ABR"], values="ARR_ON_TIME", aggfunc="mean")
df.pivot_table(index=["OP_CARRIER"],columns=["DEST_STATE_ABR"], values="ARR_ON_TIME", aggfunc="mean")

DEST_STATE_ABR,AZ,CA,CO,FL,GA,IL,LA,MA,MD,MI,MN,MO,NC,NJ,NV,NY,PA,SC,SD,TN,TX,VA,WA
OP_CARRIER,,,,,,,,,,,,,,,,,,,,,,,
9E,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.948276,NaN,0.920455,0.938596,NaN,NaN,NaN,NaN,0.886477,NaN,NaN,NaN,NaN,NaN,NaN,NaN
AA,0.911458,NaN,NaN,0.889680,NaN,0.742857,NaN,NaN,NaN,NaN,NaN,NaN,0.894942,NaN,NaN,NaN,0.893082,NaN,NaN,NaN,0.817434,NaN,NaN
AS,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.756494
B6,NaN,NaN,NaN,0.821782,NaN,NaN,NaN,0.834802,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
DL,NaN,NaN,NaN,NaN,0.929342,NaN,NaN,NaN,NaN,NaN,0.913580,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
F9,NaN,NaN,0.822785,0.670886,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
G4,0.775000,NaN,NaN,0.762546,0.622449,NaN,0.555556,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.732620,0.666667,0.763441,0.750000,0.714286,NaN
MQ,NaN,NaN,NaN,0.758364,NaN,0.855422,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.774648,NaN,NaN
NK,NaN,0.760479,NaN,0.754928,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.804979,NaN,NaN,0.781818,NaN,NaN,NaN,NaN,NaN


*The airlines serve different destinations. It could be that congestion at destination (possibly also affecting incoming flight) is responsible for the delays. We cannot exclude Simpson's type paradoxes when comparing these airlines on aggregate metrics.*

### Section 5: Causes of Delay

**Task 5.1**: What are the most commonly cited reasons for flight delays by the airlines?

In [48]:
delay_cause_cols = ["CARRIER_DELAY", "WEATHER_DELAY", "NAS_DELAY", 
                    "SECURITY_DELAY", "LATE_AIRCRAFT_DELAY"]

In [25]:
df[delay_cause_cols].sum()

CARRIER_DELAY          109273.0
WEATHER_DELAY           21687.0
NAS_DELAY               61347.0
SECURITY_DELAY            781.0
LATE_AIRCRAFT_DELAY    108638.0
dtype: float64

*Carrier-related delays (include late inbound aircraft) are far more common than air traffic, weather, or security delays.*

### Optional Section 6: (Open Ended) Ask yourself a question, not asked above, that you find interesting and answer it.

In [14]:
df["DEP_DELAY_HH"]= (df["CRS_DEP_TIME"] - df["DEP_TIME"])

In [18]:
col_names = list(df.columns)

In [19]:
col_names.insert(0,col_names.pop())

In [20]:
col_names

['DEP_DELAY_HH',
 'YEAR',
 'MONTH',
 'DAY_OF_MONTH',
 'DAY_OF_WEEK',
 'FL_DATE',
 'OP_CARRIER',
 'OP_CARRIER_FL_NUM',
 'ORIGIN',
 'ORIGIN_CITY_NAME',
 'ORIGIN_STATE_ABR',
 'DEST',
 'DEST_CITY_NAME',
 'DEST_STATE_ABR',
 'CRS_DEP_TIME',
 'DEP_TIME',
 'DEP_DELAY',
 'DEP_DELAY_NEW',
 'DEP_DEL15',
 'DEP_TIME_BLK',
 'CRS_ARR_TIME',
 'ARR_TIME',
 'ARR_DELAY',
 'ARR_DELAY_NEW',
 'ARR_DEL15',
 'ARR_TIME_BLK',
 'CANCELLED',
 'CANCELLATION_CODE',
 'DIVERTED',
 'CARRIER_DELAY',
 'WEATHER_DELAY',
 'NAS_DELAY',
 'SECURITY_DELAY',
 'LATE_AIRCRAFT_DELAY']

In [21]:
df = df.reindex(columns = col_names)

In [22]:
df

,DEP_DELAY_HH,YEAR,MONTH,DAY_OF_MONTH,DAY_OF_WEEK,FL_DATE,OP_CARRIER,OP_CARRIER_FL_NUM,ORIGIN,ORIGIN_CITY_NAME,ORIGIN_STATE_ABR,DEST,DEST_CITY_NAME,DEST_STATE_ABR,CRS_DEP_TIME,DEP_TIME,DEP_DELAY,DEP_DELAY_NEW,DEP_DEL15,DEP_TIME_BLK,CRS_ARR_TIME,ARR_TIME,ARR_DELAY,ARR_DELAY_NEW,ARR_DEL15,ARR_TIME_BLK,CANCELLED,CANCELLATION_CODE,DIVERTED,CARRIER_DELAY,WEATHER_DELAY,NAS_DELAY,SECURITY_DELAY,LATE_AIRCRAFT_DELAY
0,1.0,2021,1,10,7,1/10/2021 12:00:00 AM,OO,3878,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,1901,1900.0,-1.0,0.0,0.0,1900-1959,2025,2023.0,-2.0,0.0,0.0,2000-2059,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
1,44.0,2021,1,4,1,1/4/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,756.0,-4.0,0.0,0.0,0800-0859,921,904.0,-17.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
2,46.0,2021,1,3,7,1/3/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,754.0,-6.0,0.0,0.0,0800-0859,921,903.0,-18.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
3,2.0,2021,1,3,7,1/3/2021 12:00:00 AM,OO,3989,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,1245,1243.0,-2.0,0.0,0.0,1200-1259,1403,1339.0,-24.0,0.0,0.0,1400-1459,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
4,-7.0,2021,1,11,1,1/11/2021 12:00:00 AM,OO,3934,PIT,"Pittsburgh, PA",PA,MSP,"Minneapolis, MN",MN,800,807.0,7.0,7.0,0.0,0800-0859,930,913.0,-17.0,0.0,0.0,0900-0959,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32615,-6.0,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2407,PIT,"Pittsburgh, PA",PA,IAH,"Houston, TX",TX,1639,1645.0,6.0,6.0,0.0,1600-1659,1859,1853.0,-6.0,0.0,0.0,1800-1859,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32616,5.0,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2109,PIT,"Pittsburgh, PA",PA,DEN,"Denver, CO",CO,838,833.0,-5.0,0.0,0.0,0800-0859,1022,1019.0,-3.0,0.0,0.0,1000-1059,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32617,6.0,2021,12,20,1,12/20/2021 12:00:00 AM,UA,2108,PIT,"Pittsburgh, PA",PA,IAH,"Houston, TX",TX,610,604.0,-6.0,0.0,0.0,0600-0659,835,822.0,-13.0,0.0,0.0,0800-0859,0.0,NaN,0.0,NaN,NaN,NaN,NaN,NaN
32618,-104.0,2021,12,20,1,12/20/2021 12:00:00 AM,UA,1810,PIT,"Pittsburgh, PA",PA,DEN,"Denver, CO",CO,1600,1704.0,64.0,64.0,1.0,1600-1659,1744,1841.0,57.0,57.0,1.0,1700-1759,0.0,NaN,0.0,57.0,0.0,0.0,0.0,0.0


---
## END